homeassistant/components/homekit_controller/fan.py

```diff
@@ -0,0 +1,254 @@
+        """Initialise the fan."""
+        self._on = None
+        self._features = 0
+        self._rotation_direction = 0
+        self._rotation_speed = 0
+        self._swing_mode = 0
+
+        super().__init__(*args)
+
+    def get_characteristic_types(self):
+        """Define the homekit characteristics the entity cares about."""
+        return [
+            CharacteristicsTypes.SWING_MODE,
+            CharacteristicsTypes.ROTATION_DIRECTION,
+            CharacteristicsTypes.ROTATION_SPEED,
+        ]
+
+    def _setup_rotation_direction(self, char):
+        self._features |= SUPPORT_DIRECTION
+
+    def _setup_rotation_speed(self, char):
+        self._features |= SUPPORT_SET_SPEED
+
+    def _setup_swing_mode(self, char):
+        self._features |= SUPPORT_OSCILLATE
+
+    def _update_rotation_direction(self, value):
+        self._rotation_direction = value
+
+    def _update_rotation_speed(self, value):
+        self._rotation_speed = value
+
+    def _update_swing_mode(self, value):
+        self._swing_mode = value
+
+    @property
+    def is_on(self):
+        """Return true if device is on."""
+        return self._on
+
+    @property
+    def speed(self):
+        """Return the current speed."""
+        if not self.is_on:
+            return SPEED_OFF
+        if self._rotation_speed > SPEED_TO_PCNT[SPEED_MEDIUM]:
+            return SPEED_HIGH
+        if self._rotation_speed > SPEED_TO_PCNT[SPEED_LOW]:
+            return SPEED_MEDIUM
+        if self._rotation_speed > SPEED_TO_PCNT[SPEED_OFF]:
+            return SPEED_LOW
+        return SPEED_OFF
+
+    @property
+    def speed_list(self):
+        """Get the list of available speeds."""
+        if self.supported_features & SUPPORT_SET_SPEED:
+            return [SPEED_OFF, SPEED_LOW, SPEED_MEDIUM, SPEED_HIGH]
+        return []
+
+    @property
+    def current_direction(self):
+        """Return the current direction of the fan."""
+        return HK_DIRECTION_TO_HA[self._rotation_direction]
+
+    @property
+    def oscillating(self):
+        """Return whether or not the fan is currently oscillating."""
+        return self._swing_mode == 1
+
+    @property
+    def supported_features(self):
+        """Flag supported features."""
+        return self._features
+
+    async def async_set_direction(self, direction):
+        """Set the direction of the fan."""
+        if self.supported_features & SUPPORT_DIRECTION:
+            await self._accessory.put_characteristics(
+                [
+                    {
+                        "aid": self._aid,
+                        "iid": self._chars["rotation.direction"],
+                        "value": DIRECTION_TO_HK[direction],
+                    }
+                ]
+            )
+
+    async def async_set_speed(self, speed):
+        """Set the speed of the fan."""
+        if speed == SPEED_OFF:
+            return await self.async_turn_off()
+
+        if self.supported_features & SUPPORT_SET_SPEED:
```

Comment:
A bit off topic:
This should preferably be checked by the base entity integration by submitting supported feature value when registering the service.

It's currently missing in the registration:
https://github.com/home-assistant/home-assistant/blob/f400b778371f032e26e3192870830719f16d96a2/homeassistant/components/fan/__init__.py#L81-L83